[AIRFLOW-3605] Load plugins from entry_points

drewsonne opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. 
  - https://issues.apache.org/jira/browse/AIRFLOW-3605

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Added logic to load AirflowPlugins from the [entry_points](https://setuptools.readthedocs.io/en/latest/pkg_resources.html#id16). Rather than moving files on the OS, this allows plugins to be installed and distributed via `pip`. 

Also added a callback to execute business logic from the plugin when the plugin is loaded.

### Tests

- [x] My PR adds the following unit tests:
 - `plugins_manager.PluginsEntrypointLoad`
 - `plugins_manager.PluginsTestValidator`

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

~I haven't written the documentation yet, as I wanted to see if there was an appetite for this kind of loading of plugins.~

### Code Quality

- [x] Passes `flake8`

drewsonne committed:
[AIRFLOW-3605] Add entrypoint plugin docs

This documentation came from https://github.com/apache/incubator-airflow/pull/730 which had already started work on a PR for this functionality.

ashb commented on airflow/plugins_manager.py:
```diff
@@ -90,19 +142,17 @@ def validate(cls):
             import_errors[filepath] = str(e)
 
+plugins += load_entrypoint_plugins(
+    pkg_resources.iter_entry_points('airflow.plugins'))
```
Should the currently loaded plugins be passed in to `load_entrypoint_plugs()`? I.e something like:

```python
plugins += load_entrypoint_plugins(
    pkg_resources.iter_entry_points('airflow.plugins'),
    plugins)
```

drewsonne replied:
You are correct, we're not checking for double loading of plugins from both sources. Will change.

ashb commented on docs/plugins.rst:
```diff
@@ -220,3 +220,26 @@ Note on role based views
+            ]
+        }
+    )
+
```
Could you extend this doc to make it clear under what name this plugin will be available? Because I don't think that _any_ of these names actually appears in the airflow plugin import statement at use time? (Though that might be clearer if it did?)

drewsonne replied:
I'll expand the documentation when I get home, but given a class like

```python
# my_package/my_plugin.py
from airflow.plugins_manager import AirflowPlugin

class MyAirflowPlugin(AirflowPlugin):
    name = 'my_namespace.MyPlugin'
```
I think (I can't remember off the top of my head, but I will test) that the plugin would be exposed under `airflow.operators.my_namespace.MyPlugin`, `airflow.sensors.my_namespace.MyPluigin`, etc.

drewsonne replied:
I'd write some test cases for this functionality, but it is a bit difficult given the structure of this file. Would there be objection to wrapping this functionality in some functions as I did with `is_valid_plugin(...)` to make this a bit easier to test?

drewsonne replied:
Before I spent time working on testcases, this would be the idea (and I'd put it in a separate PR) https://github.com/drewsonne/incubator-airflow/pull/1/files

ashb replied:
The class as you have it in that PR1 you link to looks fine from a very quick glance.

I was more thinking that in the setup.py example in the docs that the `my_plugin` (and maybe `MyAirflowPlugin`?) name doesn't matter - it's not part of the the "name" that is exposed by airflow is it? It would be `from airflow.operators.my_namespace import MyOperator` right?

drewsonne replied:
Ah right, i have not used the “name” of the entry point as it is already handled by the ‘AirflowPlugin.name’. We could use it, but it seems like a precedence would just complicate things.

drewsonne replied:
And yes, that import is correct. It's using the same namespace building logic as the file based plugins, so it shouldn't change beyond that.

drewsonne committed:
[AIRFLOW-3605] Extend plugin loading functionality

Added business logic to import AirflowPlugin classes through entry_points.
This means we don’t have to interact with the file system directly to install plugins, and can manage them via `pip`.

ashb reviewed (approved):
Code looks good, I haven't looked at why tests are failing thought :)

feng-tao commented on tests/plugins/test_plugins_manager_www.py:
```diff
@@ -83,3 +85,66 @@ def test_menu_links(self):
         self.assertEqual('Test Menu Link', menu_link.name)
+
+
+class MockPluginA(AirflowPlugin):
```
sorry, but could you move the MockPlugin to test_plugin file which is a centralized place for all the test plugins?

drewsonne replied:
Done, I just have to wait for the tests to run again

feng-tao commented on tests/plugins/test_plugins_manager_www.py:
```diff
@@ -83,3 +85,66 @@ def test_menu_links(self):
+    name = 'plugin-c'
+
+
+class PluginsTestEntrypointLoad(unittest.TestCase):
```
are these two unit test only worked for old UI? If not I would suggest to move to test_plugins_manager_rbac.py once we delete the old UI code, we will remove this unit test file as well.

drewsonne replied:
I'm not sure, as I haven't touched on the UI parts of the code base yet.

This code is not really anything to do with the UI as far as I'm aware. Or at least, it's not specific to one UI or the other. When I pulled the above PR changes in, this is where my code was move to, so the old test/plugin_manager.py file was renamed to this.

tl;dr it's not UI specific(?), so I'm not sure where it goes.

feng-tao replied:
I guess we could leave it as it is now. But FYI, the old UI and new UI depends on different flask framework which works a bit differently with the plugin manager(e.g. old UI could integrate with blueprint while old UI could not etc). But tldr: let's leave it as it is now.

drewsonne replied:
From what I've seen so far, the plugin manager loads the plugins, and dumps their respective components (hooks, operators, blueprints, etc) into lists and then it's up to other components to decide which lists to reference.

I'm happy to put in another PR to move it so keep thing Clean™, but I'm not sure where it'd go. Are there plans for the old UI to be deprecated at a time in the future? That could solve the problem.

feng-tao reviewed (approved):
Two small nits, otherwise lgtm
